src/openzeppelin/utils/presets/UniversalDeployer.cairo

```diff
@@ -0,0 +1,72 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts for Cairo v0.4.0 (utils/presets/UniversalDeployer.cairo)
+
+%lang starknet
+
+from starkware.starknet.common.syscalls import get_caller_address, deploy
+from starkware.cairo.common.cairo_builtins import HashBuiltin
+from starkware.cairo.common.hash import hash2
+from starkware.cairo.common.bool import FALSE, TRUE
+
+@event
+func ContractDeployed(
+    address: felt,
+    deployer: felt,
+    unique: felt,
+    classHash: felt,
+    calldata_len: felt,
+    calldata: felt*,
+    salt: felt
+) {
+}
+
+@external
+func deployContract{
+    syscall_ptr: felt*,
+    pedersen_ptr: HashBuiltin*,
+    range_check_ptr
+}(
+    classHash: felt,
+    salt: felt,
+    unique: felt,
+    calldata_len: felt,
+    calldata: felt*
+) -> (address: felt) {
+    alloc_locals;
+    let (deployer) = get_caller_address();
+
+    local _salt;
+    local from_zero;
+    if (unique == TRUE) {
+        let (unique_salt) = hash2{hash_ptr=pedersen_ptr}(deployer, salt);
+        _salt = unique_salt;
+        from_zero = FALSE;
+        tempvar _pedersen = pedersen_ptr;
+    } else {
+        _salt = salt;
+        from_zero = TRUE;
+        tempvar _pedersen = pedersen_ptr;
+    }
+
+    tempvar pedersen_ptr = _pedersen;
```

Comment:
I think we don't need to increment ap here with tempvar when a reference is enough:
```suggestion
    let pedersen_ptr = _pedersen;
```

Actually, the reference shouldn't even be necessary, from my understanding, if you rename `_perdersen` to `pedersen_ptr` rebinding the reference (because in both branches we should have [ap - 1]). But, for some reason, when doing this, the line emitting the event seems to revoke the reference. This is an expected behavior @bbrandtom ?

Reply:
I mean, this works (as I expect):

```
    if (unique == TRUE) {
        let (unique_salt) = hash2{hash_ptr=pedersen_ptr}(deployer, salt);
        _salt = unique_salt;
        from_zero = FALSE;
        tempvar pedersen_ptr = pedersen_ptr;
    } else {
        _salt = salt;
        from_zero = TRUE;
        tempvar pedersen_ptr = pedersen_ptr;
    }

    let (address) = deploy(
        class_hash=classHash,
        contract_address_salt=_salt,
        constructor_calldata_size=calldata_len,
        constructor_calldata=calldata,
        deploy_from_zero=from_zero,
    );

    return (address=address)
```

While this doesn't (I would expect it to work too):

```
    if (unique == TRUE) {
        let (unique_salt) = hash2{hash_ptr=pedersen_ptr}(deployer, salt);
        _salt = unique_salt;
        from_zero = FALSE;
        tempvar pedersen_ptr = pedersen_ptr;
    } else {
        _salt = salt;
        from_zero = TRUE;
        tempvar pedersen_ptr = pedersen_ptr;
    }

    let (address) = deploy(
        class_hash=classHash,
        contract_address_salt=_salt,
        constructor_calldata_size=calldata_len,
        constructor_calldata=calldata,
        deploy_from_zero=from_zero,
    );


    ContractDeployed.emit(
        address=address,
        deployer=deployer,
        unique=unique,
        classHash=classHash,
        calldata_len=calldata_len,
        calldata=calldata,
        salt=salt
    );

    return (address=address)
```

Reply:
@ericnordelo what do you mean by it doesn't work? what exact error are you getting? and i'm not sure the point you're trying to make by removing the event emission

Reply:
The point is that the event emission is revoking the pedersen_ptr reference in the example I'm presenting when I think it shouldn't. And when you rebind the reference to the same ap reference, it works again (with the event). This looks like a potential compiler issue.

Reply:
@ericnordelo What's the rebinding that works with the event? My guess is that the revocation happens due to an arbitrary change in ap caused by a function call. Usually (after 0.5.0 IIRC), such revocations are solved implicitly by adding a local variable definition prior to the call. However, here the function itself is generated by the compiler (emit is replaced by a code using the emit_event syscall) so it might be missed.

Reply:
Hi @ArielElp, the rebinding that works with the event is the one after the `else` block in this example:

```
    (...)
    if (unique == TRUE) {
        let (unique_salt) = hash2{hash_ptr=pedersen_ptr}(deployer, salt);
        _salt = unique_salt;
        from_zero = FALSE;
        tempvar pedersen_ptr = pedersen_ptr;
    } else {
        _salt = salt;
        from_zero = TRUE;
        tempvar pedersen_ptr = pedersen_ptr;
    }
    
    // This is the rebinding:
    let pedersen_ptr = pedersen_ptr;

    let (address) = deploy(
        class_hash=classHash,
        contract_address_salt=_salt,
        constructor_calldata_size=calldata_len,
        constructor_calldata=calldata,
        deploy_from_zero=from_zero,
    );

    ContractDeployed.emit(
        address=address,
        deployer=deployer,
        unique=unique,
        classHash=classHash,
        calldata_len=calldata_len,
        calldata=calldata,
        salt=salt
    );

    return (address=address)
```

If you remove that rebinding, the reference is revoked by the `emit` call. My question is, why is it not being revoked when we just rebind the reference from ap to itself? (the same reference to the exact ap location)